libs/platform/ebpf_object.c

```diff
@@ -22,59 +22,137 @@ static ebpf_lock_t _ebpf_object_tracking_list_lock = {0};
  * 2) A pinning table entry holds a reference on it.
  * 3) A link object can hold a reference when it is associated with a hook.
  *
- * A link object can have a reference due to on of the following:
+ * A link object can have a reference due to one of the following:
  * 1) An open handle holds a reference on it.
  * 2) A pinning table entry holds a reference on it.
  *
+ * Libbpf has APIs like bpf_prog_get_next_id, bpf_map_get_next_id,
+ * and bpf_link_get_next_id to enumerate object IDs so every
+ * object gets a unique ID upon creation.
  */
-static _Requires_lock_held_(&_ebpf_object_tracking_list_lock) ebpf_list_entry_t _ebpf_object_tracking_list;
 
-static void
-_ebpf_object_tracking_list_insert(ebpf_object_t* object)
+typedef struct _ebpf_id_entry
+{
+    // Counter incremented each time a new object is stored here.
+    uint16_t counter;
+
+    // Pointer to object.
+    ebpf_object_t* object;
+} ebpf_id_entry_t;
+
+// Currently we allow a maximum of 1024 objects (links, maps,
+// and programs combined).  This can be increased in the future
+// if needed by making each array element store a linked list of
+// entries in order by ID, rather than a single entry, in which
+// case this becomes a hash table.
+static _Requires_lock_held_(&_ebpf_object_tracking_list_lock) ebpf_id_entry_t _ebpf_id_table[1024];
+
+// Get the ID last stored at a given index.
+static inline ebpf_id_t
+_get_id_from_index(uint32_t index)
+{
+    // Put the index in the high 2 bytes and the counter in the low 2 bytes.
+    // This allows us to more easily detect stale IDs while
+    // still keeping IDs in order by index.
+    return (index << 16) | _ebpf_id_table[index].counter;
+}
+
+// Get the index at which a given ID is stored.
+static inline ebpf_result_t
+_get_index_from_id(ebpf_id_t id, _Out_ uint32_t* index)
 {
+    uint32_t possible_index = id >> 16;
+    if (id != _get_id_from_index(possible_index)) {
+        return EBPF_KEY_NOT_FOUND;
+    }
+
+    *index = possible_index;
+    return EBPF_SUCCESS;
+}
+
+static ebpf_result_t
+_ebpf_object_tracking_list_insert(_Inout_ ebpf_object_t* object)
+{
+    int new_index;
+    ebpf_result_t return_value;
     ebpf_lock_state_t state;
     state = ebpf_lock_lock(&_ebpf_object_tracking_list_lock);
-    ebpf_list_insert_tail(&_ebpf_object_tracking_list, &object->global_list_entry);
+    for (new_index = 1; new_index < EBPF_COUNT_OF(_ebpf_id_table); new_index++) {
```

Comment:
Do we not use index 0?

Reply:
correct, I believe fd of 0 isn't supported in the ebpf ISA. not sure but wanted to be safe and distinguish it from unset fields in structs initialized to zero.